File: pyhexz/src/pyhexz/board.py
"""Pure Python implementation of the Flagz board game."""

import numpy as np


def _init_neighbors_map():
    """Returns a dict mapping all valid (r, c) indices to their neighbor indices.

    The neighbor indices are represented as (row, column) tuples."""

    def _valid_idx(r_c):
        r, c = r_c
        return r >= 0 and r < 11 and c >= 0 and c < 10 - r % 2

    result = {}
    for r in range(11):
        shift = r % 2  # Depending on the row, neighbors below and above are shifted.
        for c in range(10 - r % 2):
            ns = filter(
                _valid_idx,
                [
                    (r, c + 1),
                    (r - 1, c + shift),
                    (r - 1, c - 1 + shift),
                    (r, c - 1),
                    (r + 1, c - 1 + shift),
                    (r + 1, c + shift),
                ],
            )
            nr, nc = zip(*ns)  # unzip
            result[(r, c)] = (np.array(nr), np.array(nc))

    return result


class Board:
    """Numpy representation of a hexz board.

    A board is represented by an (11, 11, 10) numpy array. Each 11x10 channel is
    a one-hot encoding of the presence of specific type of piece/obstacle/etc.
    The channels are:

    * 0: flags by P1
    * 1: cell value 1-5 for P1
    * 2: cells blocked for P1 (any occupied cell or a cell next to a 5)
    * 3: next value for P1
    * 4: number of remaining flags for P1
    * 5: flags by P2
    * 6: cell value 1-5 for P2
    * 7: cells blocked for P2
    * 8: next value for P2
    * 9: number of remaining flags for P1
    * 10: grass cells with value 1-5

    An action is specified by a (2, 11, 10) numpy array. The first 11x10 channel
    represents a flag move, the second one represents a regular cell move. A
    flag move must have a single 1 set, a normal move must have a single value
    1-5 set.
    """

    shape = (11, 11, 10)

    # Used to quickly get the indices of neighbor cells.
    neighbors = _init_neighbors_map()

    def __init__(self, other=None, dtype=np.float32, board=None):
        """Generates a new randomly initialized board or returns a copy of other, if set."""
        if other is not None:
            self.b = other.b.copy()
            return
        if board is not None:
            self.b = board
            return
        self.b = np.zeros(self.shape, dtype=dtype)
        # 3 flags for each player
        self.b[4] = 3
        self.b[9] = 3
        # Even rows have 10 cells, odd rows only 9, so mark the last cell in odd rows as blocked for P1+P2.
        self.b[2, [1, 3, 5, 7, 9], 9] = 1
        self.b[7, [1, 3, 5, 7, 9], 9] = 1
        # 2-tuple of valid indices in each slice.
        free_cells = (1 - self.b[2]).nonzero()
        # 15 randomly placed stones.
        rng = np.random.default_rng()
        stones = rng.choice(np.arange(0, len(free_cells[0])), replace=False, size=15)
        self.b[2, free_cells[0][stones], free_cells[1][stones]] = 1
        self.b[7, free_cells[0][stones], free_cells[1][stones]] = 1
        free_cells = (1 - self.b[2]).nonzero()
        # 5 grass cells
        grass = rng.choice(np.arange(0, len(free_cells[0])), replace=False, size=5)
        self.b[10, free_cells[0][grass], free_cells[1][grass]] = [1, 2, 3, 4, 5]
        self.b[2, free_cells[0][grass], free_cells[1][grass]] = 1
        self.b[7, free_cells[0][grass], free_cells[1][grass]] = 1

    @classmethod
    def from_numpy(cls, board):
        if board.shape != cls.shape:
            raise ValueError("Invalid board shape: " + str(board.shape))
        return cls(board=board)

    def b_for(self, player):
        """Returns the underlying ndarray representing the board, oriented for the given player.
        Only returns a copy if the board is not already oriented for the given player.
        """
        if player == 0:
            return self.b
        b = self.b.copy()
        b[0:5], b[5:10] = self.b[5:10], self.b[0:5]
        return b

    def score(self):
        """Returns the current score as a 2-tuple."""
        return (self.b[1].sum(), self.b[6].sum())

    def result(self):
        """Returns the final result of the board.

        1 (player 0 won), 0 (draw), -1 (player 1 won).
        """
        p0, p1 = self.score()
        return np.sign(p0 - p1)

    def flags(self, player):
        return int(self.b[4 + player * 5, 0, 0])
    
    def make_move(self, player, move):
        """Makes the given move.

        Args:
          player: 0 or 1
          move: a 4-tuple of (typ, r, c, val), where typ = 0 (flag) or 1 (normal)
        Does not check that it is a valid move. Should be called only
        with moves returned from `next_moves`.
        """
        typ, r, c, val = move
        b = self.b
        b[typ + player * 5, r, c] = val
        played_flag = typ == 0
        # Block played cell for both players.
        b[2, r, c] = 1
        b[7, r, c] = 1
        # Set next value to 0 for occupied cell.
        b[3, r, c] = 0
        b[8, r, c] = 0
        # Block neighboring cells if a 5 was played.
        nx, ny = Board.neighbors[(r, c)]
        # Update next value of neighboring cells. If we played a flag, the next value is 1.
        if played_flag:
            next_val = 1
            b[4 + player * 5] -= 1
        else:
            next_val = val + 1
        if next_val <= 5:
            for nr, nc in zip(nx, ny):
                if b[2 + player * 5, nr, nc] == 0:
                    # Cell is not blocked yet.
                    if b[3 + player * 5, nr, nc] == 0:
                        b[3 + player * 5, nr, nc] = next_val
                    elif b[3 + player * 5, nr, nc] > next_val:
                        b[3 + player * 5, nr, nc] = next_val
        else:
            # Played a 5: block neighboring cells and clear next value.
            b[2 + player * 5, nx, ny] = 1
            b[3 + player * 5, nx, ny] = 0  # Clear next value.

        # Occupy neighboring grass cells.
        if not played_flag:
            self.occupy_grass(player, r, c)

    def occupy_grass(self, player, r, c):
        """Occupies the neighboring grass cells of move_idx (a 3-tuple index into a move) for player.

        Expects that the move has already been played.
        """
        nx, ny = Board.neighbors[(r, c)]
        for i, j in zip(nx, ny):
            grass_val = self.b[10, i, j]
            if grass_val > 0 and grass_val <= self.b[1 + player * 5, r, c]:
                # Occupy: first remove grass
                self.b[10, i, j] = 0
                # the rest is exactly like playing a move.
                self.make_move(player, (1, r, c, grass_val))

    def next_moves(self, player):
        """Returns all possible next moves.

        A move is represented as a (2, 11, 10) ndarray. The first slice represents
        flag moves, the second one represents normal moves. A flag move will have exactly
        one element set to 1 in slice 0. A normal move will have exactly one element set to
        1-5 in slice 1.
        """
        moves = []
        # Do we have unoccupied cells and flags left? Then we can place another one.
        if self.b[4 + player * 5, 0, 0] > 0:
            # Flag in any unoccupied cell.
            rs, cs = np.nonzero(
                self.b[2 + player * 5] == 0
            )  # funky way to get indices for all free cells.
            moves.extend((0, r, c, 1) for r, c in zip(rs, cs))
        # Collect all cells with a non-zero next value.
        rs, cs = np.nonzero(self.b[3 + player * 5])
        moves.extend((1, r, c, self.b[3 + player * 5, r, c]) for r, c in zip(rs, cs))
        return moves

    def validate(self) -> None:
        """Checks that this instance represents a valid board.

        Raises an exeption if that is not the case.
        """
        b = self.b
        if np.any(b[2, [1, 3, 5, 7, 9], 9] == 0) or np.any(b[7, [1, 3, 5, 7, 9], 9] == 0) :
            raise ValueError("Invalid cells are not marked as blocked")
        if np.any(b[1] * b[6]):
            raise ValueError("Both players have a value in the same cell.")
        if np.any(b[0] * b[5]):
            raise ValueError("Both players have a flag in the same cell.")
        if np.any(b[2] * (b[0] + b[1]) != (b[0] + b[1])):
            raise ValueError("Not all occupied cells are marked as blocked for P1.")
        if np.any(b[7] * (b[5] + b[6]) != (b[5] + b[6])):
            raise ValueError("Not all occupied cells are marked as blocked for P2.")
        if np.any(b[0] * b[1]):
            raise ValueError("P1 has a value in a cell with a flag.")
        if np.any(b[5] * b[6]):
            raise ValueError("P2 has a value in a cell with a flag.")
        if np.any(b[3] * (b[0] + b[1] + b[2])):
            raise ValueError("P1 has a next value in an occupied cell.")
        if np.any(b[8] * (b[5] + b[6] + b[7])):
            raise ValueError("P2 has a next value in a blocked cell.")
        grass_values, grass_counts = np.unique(b[10], return_counts=True)
        if not (set(grass_values) <= set([0, 1, 2, 3, 4, 5])):
            raise ValueError(f"Wrong values for grass cells: {grass_values}")
        if len(grass_counts) > 1 and grass_counts[1:].max() > 1:
            raise ValueError(f"Duplicate grass cells: {grass_values}, {grass_counts}")
        if b.max() > 5:
            raise ValueError("Max value is > 5.")
        if b.min() < 0:
            raise ValueError("Min value is < 0.")

File: cpp/board.cc
#include "board.h"

#include <absl/log/absl_check.h>
#include <absl/status/statusor.h>
#include <absl/strings/str_cat.h>
#include <absl/strings/str_format.h>
#include <torch/torch.h>

#include <cstdlib>
#include <iterator>
#include <random>
#include <sstream>
#include <string>
#include <unordered_map>

#include "base.h"
#include "perfm.h"

#define CHANNELS_PER_PLAYER 5

// Accessor macros for the different channels.
#define I_FLAG(p) (0 + CHANNELS_PER_PLAYER * p)
#define I_VALUE(p) (1 + CHANNELS_PER_PLAYER * p)
#define I_BLOCKED(p) (2 + CHANNELS_PER_PLAYER * p)
#define I_NEXTVAL(p) (3 + CHANNELS_PER_PLAYER * p)
#define I_NFLAGS(p) (4 + CHANNELS_PER_PLAYER * p)
#define I_GRASS 10

namespace hexz {

using internal::Idx;

namespace internal {
// InitializeNeighborsMap returns a map that yields the valid neighbor cell
// indices for all valid indices of a hexz board.
std::unordered_map<Idx, std::vector<Idx>>* InitializeNeighborsMap() {
  auto* map = new std::unordered_map<Idx, std::vector<Idx>>();
  for (int r = 0; r < 11; r++) {
    int s = r % 2;
    for (int c = 0; c < 10 - s; c++) {
      Idx k{r, c};
      std::vector<Idx> ns{
          Idx{r, c + 1}, Idx{r - 1, c + s},     Idx{r - 1, c - 1 + s},
          Idx{r, c - 1}, Idx{r + 1, c - 1 + s}, Idx{r + 1, c + s},
      };
      for (const auto& n : ns) {
        if (n.IsValid()) {
          (*map)[k].push_back(n);
        }
      }
    }
  }
  return map;
}

// NeighborsOf returns the valid neighbor indices of k.
const std::vector<Idx>& NeighborsOf(const Idx& k) {
  static const std::unordered_map<Idx, std::vector<Idx>>* map =
      InitializeNeighborsMap();
  return (*map).at(k);
}

}  // namespace internal

Board::Board() {
  b_ = torch::zeros({11, 11, 10}, torch::dtype(torch::kFloat32));
}

// Copy c'tor.
Board::Board(const Board& other) { b_ = other.b_.clone(); }

Board Board::RandomBoard() {
  Board b;
  b.b_.index_put_({I_NFLAGS(0)}, 3.0f);
  b.b_.index_put_({I_NFLAGS(1)}, 3.0f);
  // Even rows have 10 cells, odd rows only 9, so mark the last cell in odd
  // rows as blocked for P1+P2.
  for (int i = 1; i <= 9; i += 2) {
    b.b_.index_put_({I_BLOCKED(0), i, 9}, 1);
    b.b_.index_put_({I_BLOCKED(1), i, 9}, 1);
  }
  // 15 randomly placed stones.
  std::uniform_int_distribution<> rnd_row(0, 10);
  std::uniform_int_distribution<> rnd_col(0, 9);
  for (int n_stones = 0; n_stones < 15;) {
    int r = rnd_row(internal::rng);
    int c = rnd_col(internal::rng);
    if (b.b_.index({I_BLOCKED(0), r, c}).item<float>() != 0) {
      continue;  // Already occupied.
    }
    b.b_.index_put_({I_BLOCKED(0), r, c}, 1);
    b.b_.index_put_({I_BLOCKED(1), r, c}, 1);
    n_stones++;
  }
  // 5 randomly placed grass cells.
  for (int n_grass = 0; n_grass < 5;) {
    int r = rnd_row(internal::rng);
    int c = rnd_col(internal::rng);
    if (b.b_.index({I_BLOCKED(0), r, c}).item<float>() != 0) {
      continue;  // Already occupied.
    }
    b.b_.index_put_({I_GRASS, r, c}, n_grass + 1);
    b.b_.index_put_({I_BLOCKED(0), r, c}, 1);
    b.b_.index_put_({I_BLOCKED(1), r, c}, 1);
    n_grass++;
  }

  return b;
}

absl::StatusOr<Board> Board::FromProto(const hexzpb::Board& board) {
  const int iFlag = static_cast<int>(hexzpb::Field::FLAG);
  if (board.resources_size() != 2 ||
      board.resources(0).num_pieces_size() <= iFlag ||
      board.resources(1).num_pieces_size() <= iFlag) {
    return absl::InvalidArgumentError("Invalid resources.num_pieces shape");
  }
  Board b;
  b.b_.index_put_({I_NFLAGS(0)},
                  static_cast<float>(board.resources(0).num_pieces(iFlag)));
  b.b_.index_put_({I_NFLAGS(1)},
                  static_cast<float>(board.resources(1).num_pieces(iFlag)));
  auto a = b.b_.accessor<float, 3>();
  int r = 0;
  int c = 0;
  if (board.flat_fields_size() != 105) {
    return absl::InvalidArgumentError(
        absl::StrCat("Expecting board with exactly 105 fields, got: ",
                     board.flat_fields_size()));
  }
  for (const auto& f : board.flat_fields()) {
    if (f.owner() < 0 || f.owner() > 2) {
      // Owner is 1-based.
      return absl::InvalidArgumentError(
          absl::StrCat("Invalid owner: ", f.owner()));
    }
    int p = f.owner() - 1;
    switch (f.type()) {
      case hexzpb::Field::NORMAL:
        if (f.value() > 0) {
          a[I_VALUE(p)][r][c] = f.value();
          a[I_BLOCKED(0)][r][c] = 1;
          a[I_BLOCKED(1)][r][c] = 1;
          break;
        }
        if (f.next_val_size() > 0 && f.next_val(0) > 0) {
          a[I_NEXTVAL(0)][r][c] = f.next_val(0);
        }
        if (f.next_val_size() > 1 && f.next_val(1) > 0) {
          a[I_NEXTVAL(1)][r][c] = f.next_val(1);
        }
        if ((f.blocked() & 1) > 0) {
          a[I_BLOCKED(0)][r][c] = 1;
        }
        if ((f.blocked() & 2) > 0) {
          a[I_BLOCKED(1)][r][c] = 1;
        }
        break;
      case hexzpb::Field::FLAG:
        a[I_FLAG(p)][r][c] = 1;
        a[I_BLOCKED(0)][r][c] = 1;
        a[I_BLOCKED(1)][r][c] = 1;
        break;
      case hexzpb::Field::GRASS:
        a[I_GRASS][r][c] = f.value();
        a[I_BLOCKED(0)][r][c] = 1;
        a[I_BLOCKED(1)][r][c] = 1;
        break;
      case hexzpb::Field::ROCK:
        a[I_BLOCKED(0)][r][c] = 1;
        a[I_BLOCKED(1)][r][c] = 1;
        break;
      default:
        return absl::InvalidArgumentError(
            absl::StrCat("Board has invalid field of type ", f.type()));
    }
    c++;
    if (c == 10 - r % 2) {
      c = 0;
      r += 1;
    }
  }
  return b;
}

torch::Tensor Board::Tensor(int player) const {
  if (player == 0) {
    return b_;
  }
  // Swap channels for player 0 and player 1. Leave grass unchanged.
  // index_select returns a new Tensor that uses its own storage.
  return b_.index_select(0, torch::tensor({5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 10}));
}

std::pair<float, float> Board::Score() const {
  return {b_.index({I_VALUE(0)}).sum().item<float>(),
          b_.index({I_VALUE(1)}).sum().item<float>()};
}

float Board::Result() const {
  auto score = Score();
  if (score.first > score.second) {
    return 1.0;
  }
  if (score.first < score.second) {
    return -1.0;
  }
  return 0.0;
}

int Board::Flags(int player) const {
  // The whole NFLAGS channel has identical numbers. Just pick (0, 0).
  return b_.index({I_NFLAGS(player), 0, 0}).item<float>();
}

void Board::MakeMove(int player, const Move& move) {
  Perfm::Scope ps(Perfm::MakeMove);
  auto b_acc = b_.accessor<float, 3>();
  ABSL_DCHECK_EQ(b_acc[I_BLOCKED(player)][move.r][move.c], 0)
      << "MakeMove on blocked field";
  ABSL_DCHECK(move.typ == Move::Typ::kFlag && move.value == 1 ||
              move.typ == Move::Typ::kNormal &&
                  b_acc[I_NEXTVAL(player)][move.r][move.c] == move.value)
      << "MakeMove: wrong value: move: " << move.DebugString()
      << " board: " << DebugString();
  if (move.typ == Move::Typ::kFlag) {
    b_acc[I_FLAG(player)][move.r][move.c] = move.value;
  } else {
    // NORMAL move.
    b_acc[I_VALUE(player)][move.r][move.c] = move.value;
  }
  bool played_flag = move.typ == Move::Typ::kFlag;
  // Occupy cell for both players
  b_acc[I_BLOCKED(0)][move.r][move.c] = 1;
  b_acc[I_BLOCKED(1)][move.r][move.c] = 1;
  // Zero next value.
  b_acc[I_NEXTVAL(0)][move.r][move.c] = 0;
  b_acc[I_NEXTVAL(1)][move.r][move.c] = 0;
  float next_val = 1;
  if (played_flag) {
    ABSL_DCHECK(Flags(player) > 0)
        << "Move " << move.DebugString() << " by " << player
        << " without flags left: " << DebugString();
    // Decrement number of flags.
    b_.index_put_({I_NFLAGS(player)}, static_cast<float>(Flags(player) - 1));
  } else {
    next_val = move.value + 1;
  }
  // Update neighboring cells.
  for (const auto& nb : NeighborsOf(Idx{move.r, move.c})) {
    if (next_val <= 5) {
      if (b_acc[I_BLOCKED(player)][nb.r][nb.c] == 0) {
        // Neighbor cell is not blocked.
        if (b_acc[I_NEXTVAL(player)][nb.r][nb.c] == 0) {
          // Neighbor cell did not have a next value yet.
          b_acc[I_NEXTVAL(player)][nb.r][nb.c] = next_val;
        } else if (b_acc[I_NEXTVAL(player)][nb.r][nb.c] > next_val) {
          // Neighbor cell's value was larger: decrease.
          b_acc[I_NEXTVAL(player)][nb.r][nb.c] = next_val;
        }
      }
    } else {
      // Played a 5: block neighboring cells and clear next value.
      b_acc[I_BLOCKED(player)][nb.r][nb.c] = 1;
      b_acc[I_NEXTVAL(player)][nb.r][nb.c] = 0;
    }
  }
  if (!played_flag) {
    // Occupy neighboring grass fields.
    for (const auto& nb : NeighborsOf(Idx{move.r, move.c})) {
      float grass_val = b_acc[I_GRASS][nb.r][nb.c];
      if (grass_val > 0 &&
          grass_val <= b_acc[I_VALUE(player)][move.r][move.c]) {
        // Occupy grass cell: remove grass value and add it as player's value.
        b_acc[I_GRASS][nb.r][nb.c] = 0;
        b_acc[I_BLOCKED(player)][nb.r][nb.c] = 0;
        b_acc[I_NEXTVAL(player)][nb.r][nb.c] = grass_val;
        MakeMove(player, Move{Move::Typ::kNormal, nb.r, nb.c, grass_val});
      }
    }
  }
}

std::vector<Move> Board::NextMoves(int player) const {
  Perfm::Scope ps(Perfm::NextMoves);
  std::vector<Move> moves;
  auto b_acc = b_.accessor<float, 3>();
  bool flag = Flags(player) > 0;
  // Flag in any unoccupied cell.
  for (int r = 0; r < 11; r++) {
    int cols = 10 - r % 2;
    for (int c = 0; c < cols; c++) {
      if (flag && b_acc[I_BLOCKED(player)][r][c] == 0) {
        moves.push_back(Move::Flag(r, c));
      }
      float next_val = b_acc[I_NEXTVAL(player)][r][c];
      if (next_val > 0) {
        moves.push_back(Move{Move::Typ::kNormal, r, c, next_val});
      }
    }
  }
  return moves;
}

float Board::CellValue(int player, Channel ch, int r, int c) const {
  int ch_idx =
      static_cast<int>(ch == kGrass ? ch : ch + CHANNELS_PER_PLAYER * player);
  return b_.index({ch_idx, r, c}).item<float>();
}

void Board::SetCellValue(int player, Channel ch, int r, int c, float value) {
  int ch_idx =
      static_cast<int>(ch == kGrass ? ch : ch + CHANNELS_PER_PLAYER * player);
  b_.index_put_({ch_idx, r, c}, value);
}

void Board::SetRemainingFlags(int player, int n_flags) {
  b_.index_put_({I_NFLAGS(player)}, static_cast<float>(n_flags));
}

std::string Board::ShortDebugString() const {
  const auto [s0, s1] = Score();
  return absl::StrFormat("Board(flags: %d-%d score: %.0f-%.0f)", Flags(0),
                         Flags(1), s0, s1);
}

std::string Board::DebugString() const {
  std::ostringstream os;
  os << "Board(\n";
  os << "  flags:(" << Flags(0) << ", " << Flags(1) << ")\n";
  os << "  score:" << Score() << "\n";
  os << "  fields: [\n";
  auto b_acc = b_.accessor<float, 3>();
  for (int r = 0; r < 11; r++) {
    int cols = 10 - r % 2;
    for (int c = 0; c < cols; c++) {
      // Output row/col indices and values of all 11 channels
      // Example row: (0, 1): 0 0 1 0 0 1 1 1 0
      os << "    (" << r << ", " << c << "): ";
      for (int i = 0; i < 11; i++) {
        int iv = static_cast<int>(b_acc[i][r][c]);
        if (i > 0) {
          os << "|";
        }
        os << iv;
      }
      if (b_acc[I_BLOCKED(0)][r][c] > 0 || b_acc[I_BLOCKED(1)][r][c] > 0) {
        os << "|b";
      }
      if (b_acc[I_FLAG(0)][r][c] > 0 || b_acc[I_FLAG(1)][r][c] > 0) {
        os << "|f";
      }
      if (b_acc[I_VALUE(0)][r][c] > 0 || b_acc[I_VALUE(1)][r][c] > 0) {
        os << "|v";
      }
      if (b_acc[I_NEXTVAL(0)][r][c] > 0 || b_acc[I_NEXTVAL(1)][r][c] > 0) {
        os << "|n";
      }
      if (b_acc[I_GRASS][r][c] > 0) {
        os << "|g";
      }
      os << "\n";
    }
  }
  os << "  ]\n";
  os << ")";
  return os.str();
}

namespace {

struct RolloutCell {
  int8_t val;
  int8_t next;
  int8_t blocked;
};

class FastGameState {
 public:
  // Initializes this state using the data from board.
  explicit FastGameState(const Board& board) {
    auto t = board.Tensor(0);
    auto acc = t.accessor<float, 3>();
    nflags[0] = acc[I_NFLAGS(0)][0][0];
    nflags[1] = acc[I_NFLAGS(1)][0][0];
    nmoves[0] = 0;
    nmoves[1] = 0;
    free_cells[0] = 0;
    free_cells[1] = 0;
    for (int r = 0; r < 11; r++) {
      for (int c = 0; c < 10 - (r & 1); c++) {
        for (int i = 0; i < 2; i++) {
          b[i][r][c].val = acc[I_VALUE(i)][r][c];
          b[i][r][c].next = acc[I_NEXTVAL(i)][r][c];
          if (b[i][r][c].next > 0) {
            nmoves[i]++;
          }
          b[i][r][c].blocked = acc[I_BLOCKED(i)][r][c];
          if (b[i][r][c].blocked == 0) {
            free_cells[i]++;
          }
        }
        grass[r][c] = acc[I_GRASS][r][c];
      }
    }
  }

  bool FastNextMove(int turn, Move& m) const {
    int moves_cnt = 0;
    float j = 0;
    bool can_place_flag = nflags[turn] > 0 && free_cells[turn] > 0;
    bool pick_flag =
        can_place_flag &&
        (nmoves[turn] == 0 ||
         internal::UnitRandom() <=
             nflags[turn] / static_cast<float>(nmoves[turn] + nflags[turn]));
    if (pick_flag) {
      int n = internal::RandomInt(0, free_cells[turn] - 1);
      int k = 0;
      for (int r = 0; r < 11; r++) {
        for (int c = 0; c < 10 - (r & 1); c++) {
          if (b[turn][r][c].blocked == 0) {
            if (k == n) {
              m = Move{Move::Typ::kFlag, r, c, 1.0};
              return true;
            }
            k++;
          }
        }
      }
    }
    if (nmoves[turn] > 0) {
      // No flag picked. Try a normal move.
      int n = internal::RandomInt(0, nmoves[turn] - 1);
      int k = 0;
      for (int r = 0; r < 11; r++) {
        for (int c = 0; c < 10 - (r & 1); c++) {
          float next_val = b[turn][r][c].next;
          if (next_val > 0) {
            if (k == n) {
              m = Move{Move::Typ::kNormal, r, c, next_val};
              return true;
            }
            k++;
          }
        }
      }
    }
    return false;
  }

  void FastMakeMove(int turn, const Move& m) {
    if (m.typ == Move::Typ::kNormal) {
      b[turn][m.r][m.c].val = m.value;
    }
    // Occupy cell for both players.
    b[0][m.r][m.c].blocked = 1;
    b[1][m.r][m.c].blocked = 1;
    free_cells[0]--;
    free_cells[1]--;
    // Update remaining moves.
    if (b[turn][m.r][m.c].next > 0) {
      nmoves[turn]--;
    }
    if (b[1 - turn][m.r][m.c].next > 0) {
      nmoves[1 - turn]--;
    }
    // Zero next value.
    b[0][m.r][m.c].next = 0;
    b[1][m.r][m.c].next = 0;
    int next_val = 1;
    if (m.typ == Move::Typ::kFlag) {
      nflags[turn]--;
    } else {
      next_val = m.value + 1;
    }
    for (const auto& nb : NeighborsOf(Idx{m.r, m.c})) {
      if (next_val <= 5) {
        if (b[turn][nb.r][nb.c].blocked == 0) {
          // Neighbor cell is not blocked.
          if (b[turn][nb.r][nb.c].next == 0) {
            // Neighbor cell did not have a next value yet.
            nmoves[turn]++;
            b[turn][nb.r][nb.c].next = next_val;
          } else if (b[turn][nb.r][nb.c].next > next_val) {
            // Neighbor cell's value was larger: decrease.
            b[turn][nb.r][nb.c].next = next_val;
          }
        }
      } else {
        // Played a 5: block neighboring cells and clear next value.
        b[turn][nb.r][nb.c].blocked = 1;
        free_cells[turn]--;
        if (b[turn][nb.r][nb.c].next > 0) {
          nmoves[turn]--;
          b[turn][nb.r][nb.c].next = 0;
        }
      }
    }
    if (m.typ == Move::Typ::kNormal) {
      for (const auto& nb : NeighborsOf(Idx{m.r, m.c})) {
        float grass_val = grass[nb.r][nb.c];
        if (grass_val > 0 && grass_val <= b[turn][m.r][m.c].val) {
          // Occupy grass cell: remove grass value and add it as player's value.
          grass[nb.r][nb.c] = 0;
          b[turn][nb.r][nb.c].blocked = 0;
          b[turn][nb.r][nb.c].next = grass_val;
          nmoves[turn]++;  // Will be removed in the recursive call.
          FastMakeMove(turn, Move{Move::Typ::kNormal, nb.r, nb.c, grass_val});
        }
      }
    }
  }

  float Result() const {
    int score[2] = {0, 0};
    for (int r = 0; r < 11; r++) {
      for (int c = 0; c < 10 - (r & 1); c++) {
        score[0] += b[0][r][c].val;
        score[1] += b[1][r][c].val;
      }
    }
    if (score[0] < score[1]) return -1;
    if (score[0] == score[1]) return 0;
    return 1;
  }

 private:
  RolloutCell b[2][11][10];
  int grass[11][10];
  int nflags[2];
  int nmoves[2];
  int free_cells[2];
};

}  // namespace

float FastRandomPlayout(int turn, const Board& board) {
  FastGameState s(board);
  int n_moves = 0;
  while (n_moves < 200) {
    // Find random next move.
    Move m;
    if (!s.FastNextMove(turn, m)) {
      turn = 1 - turn;
      if (!s.FastNextMove(turn, m)) {
        // Game over.
        return s.Result();
      }
    }
    s.FastMakeMove(turn, m);
    turn = 1 - turn;
    n_moves++;
  }
  ABSL_LOG(FATAL) << "FastRandomPlayout error: did not finish the game";
  return 0.0;  // Never reached
}

}  // namespace hexz
